File: 67-add-binary/add-binary.java
class Solution {
    public String addBinary(String a, String b) {
        int l1 = a.length()-1;
        int l2 = b.length()-1;
        StringBuilder sb = new StringBuilder();
        int carry = 0;
        
        while(l1>=0 || l2>=0){
            
            if(l1<0){
                if(b.charAt(l2)=='1'){
                    if(carry == 1){
                        sb.append("0");
                    }else{
                        sb.append("1");
                    }
                }else{
                    sb.append(carry);
                    carry = 0;
                }
            }else if(l2<0){
                if(a.charAt(l1)=='1'){
                    if(carry == 1){
                        sb.append("0");
                    }else{
                        sb.append("1");
                    }
                }else{
                    sb.append(carry);
                    carry = 0;
                }
            }else{
                if(carry == 1){
                    if(a.charAt(l1) == '1'){
                        sb.append(b.charAt(l2));
                    }else{
                        if(b.charAt(l2)== '1'){
                            sb.append("0");
                            carry = 1;
                        }else{
                            sb.append("1");
                            carry = 0;
                        }
                    }
                }else{
                    if(a.charAt(l1) == '1'){
                        if(b.charAt(l2)== '1'){
                            sb.append("0");
                            carry = 1;
                        }else{
                            sb.append("1");
                        }
                    }else{
                        sb.append(b.charAt(l2));  
                    }
                }
                
            }

            l1--;
            l2--;

        }

        if(carry==1){
                sb.append("1");
        }


        return sb.reverse().toString();

    }

   
}

File: 67-add-binary/add-binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        int m=a.length();
        int n=b.length();

        string ans="";

        if(m!=n)
        {
           while (m < n) 
           {
              a = "0" + a;
              m++;
           }

            while (n < m) {
            b = "0" + b;
    n++;
}
        }

        cout<<a<<" "<<b;

        int carry=0;
        int i=m-1;

        while(i>=0)
        {
           if((a[i]=='0' && b[i]=='1') || (a[i]=='1' && b[i]=='0'))
           {
               if(carry==0)
               {
                ans.push_back('1');
                 carry=0;
               }
               else{
                ans.push_back('0');
                carry=1;
               }
           }
           else if(a[i]=='1' && b[i]=='1'){
                if(carry==0)
                {
                    carry=1;
                    ans.push_back('0');
                }
                else{
                    carry=1;
                    ans.push_back('1');
                }
           } 

           else{
              if(carry==0)
              {
                ans.push_back('0');
              }
              else{
                ans.push_back('1');
                 carry=0;
              }
           }
            i--;
        }

        if(carry==1)
        {
             ans.push_back('1');
        }

        reverse(ans.begin(),ans.end());

        return ans;
    }
};